src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    /// referenced by `ptr` to fit `new_layout`.
+    ///
+    /// * `ptr` must have previously been provided via this allocator.
+    ///
+    /// * `layout` must *fit* the `ptr` (see above). (The `new_layout`
+    ///   argument need not fit it.)
+    ///
+    /// Behavior undefined if either of latter two constraints are unmet.
+    ///
+    /// In addition, `new_layout` should not impose a different alignment
+    /// constraint than `layout`. (In other words, `new_layout.align()`
+    /// should equal `layout.align()`.)
+    /// However, behavior is well-defined (though underspecified) when
+    /// this constraint is violated; further discussion below.
+    ///
+    /// If this returns `Ok`, then ownership of the memory block
+    /// referenced by `ptr` has been transferred to this
+    /// allocator. The memory may or may not have been freed, and
+    /// should be considered unusable (unless of course it was
+    /// transferred back to the caller again via the return value of
+    /// this method).
+    ///
+    /// Returns `Err` only if `new_layout` does not meet the allocator's
+    /// size and alignment constraints of the allocator or the
+    /// alignment of `layout`, or if reallocation otherwise fails. (Note
+    /// that did not say "if and only if" -- in particular, an
+    /// implementation of this method *can* return `Ok` if
+    /// `new_layout.align() != old_layout.align()`; or it can return `Err`
+    /// in that scenario, depending on whether this allocator
+    /// can dynamically adjust the alignment constraint for the block.)
+    ///
+    /// If this method returns `Err`, then ownership of the memory
+    /// block has not been transferred to this allocator, and the
+    /// contents of the memory block are unaltered.
+    unsafe fn realloc(&mut self,
+                      ptr: Address,
+                      layout: Layout,
+                      new_layout: Layout) -> Result<Address, AllocErr> {
+        let (min, max) = self.usable_size(&layout);
+        let s = new_layout.size();
+        // All Layout alignments are powers of two, so a comparison
+        // suffices here (rather than resorting to a `%` operation).
+        if min <= s && s <= max && new_layout.align() <= layout.align() {
+            return Ok(ptr);
+        } else {
+            let new_size = new_layout.size();
+            let old_size = layout.size();
+            let result = self.alloc(new_layout);
+            if let Ok(new_ptr) = result {
+                ptr::copy(ptr as *const u8, new_ptr, cmp::min(old_size, new_size));
+                self.dealloc(ptr, layout);
+            }
+            result
+        }
+    }
+
+    /// Behaves like `fn alloc`, but also ensures that the contents
+    /// are set to zero before being returned.
+    unsafe fn alloc_zeroed(&mut self, layout: Layout) -> Result<Address, AllocErr> {
+        let size = layout.size();
+        if size > ::core::isize::MAX as usize {
```

Comment:
because one can currently construct a `Layout` that overflows `isize`?

I suppose I could instead just say that the state of the contents are unspecified if the layout size overflows `isize`. Does that seem preferable?